pallets/settlement/src/lib.rs

```diff
@@ -1320,76 +1368,105 @@ impl<T: Trait> Module<T> {
+                                                (leg_id, FailureReason::GetMercatMediatorData)
+                                            })?
                                     }
                                     _ => {
-                                        return Err((leg_id, FailureReason::GetMercatMediatorData));
+                                        return Err((
+                                            leg_id,
+                                            FailureReason::InsufficientMercatAuthorizations,
+                                        ));
                                     }
-                                };
+                                }
+                                .map_err(|_| {
+                                    (leg_id, FailureReason::InsufficientMercatAuthorizations)
+                                })?;
 
                                 // Read the mercat_accounts from the confidential-asset pallet.
-                                if let Ok((from_mercat, from_mercat_balance)) =
-                                    Self::get_account_and_balance(
-                                        &leg.from.did,
-                                        &leg.from_account_id,
+                                let from_mercat_pending_state = Self::get_pending_state(
+                                    &leg.from.did,
+                                    &leg.from_account_id,
+                                    instruction_id,
+                                )
+                                .map_err(|_| (leg_id, FailureReason::ConfidentialTransfer))?;
+
+                                // Get receiver's current balance.
+                                let (to_mercat, to_mercat_balance) =
+                                    Self::get_account_and_balance(&leg.to.did, &leg.to_account_id)
+                                        .map_err(|_| (leg_id, FailureReason::GetMercatAccount))?;
+
+                                // Get sender's current balance.
+                                let (from_mercat, _) = Self::get_account_and_balance(
+                                    &leg.from.did,
+                                    &leg.from_account_id,
+                                )
+                                .map_err(|_| (leg_id, FailureReason::GetMercatAccount))?;
+
+                                // Verify the proofs.
+                                let mut rng = rng::Rng::default();
+                                let _result = TransactionValidator {}
+                                    .verify_transaction(
+                                        &decoded_justified_tx,
+                                        &from_mercat,
+                                        &from_mercat_pending_state,
+                                        &to_mercat,
+                                        &to_mercat_balance, // If the only reason we need this is updating balance, we should remove it.
```

Comment:
We had some debate over this a while ago. I still believe that verify should only return success or failure and should not do balance update.

Reply:
yeah, I agree, I'll bundle it with the other refactoring I'm doing on the library side.